In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()


Pada code ini digunakan untuk mempersiapkan library yang diperlukan dalam pengolahan dataset Iris. numpy digunakan untuk pengolahan data numerik, pandas digunakan untuk mengolah data dalam bentuk DataFrame, sedangkan load_iris digunakan untuk mengambil dataset Iris yang tersedia pada Scikit-Learn. Dataset Iris berisi data pengukuran bunga Iris yang nantinya digunakan sebagai data dalam proses klasifikasi menggunakan algoritma Naive Bayes.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]
# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)


     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Pada code ini digunakan untuk mengolah dataset Iris menjadi bentuk DataFrame, menambahkan label species, kemudian memisahkan data fitur dan target yang akan digunakan dalam proses klasifikasi. pd.DataFrame() digunakan untuk mengubah data Iris menjadi DataFrame dengan nama kolom yang berasal dari iris.feature_names. Selanjutnya, df["Species"] digunakan untuk menambahkan kolom species berdasarkan target pada dataset Iris. Setelah itu, X digunakan sebagai data fitur dengan menghapus kolom Species, sedangkan y digunakan sebagai target atau label yang berisi nama species. Terakhir, print(X) digunakan untuk menampilkan seluruh data fitur yang akan digunakan dalam proses klasifikasi.   

In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

Pada code y digunakan untuk menentukan target atau label yang akan diprediksi oleh model. y = df["Species"] mengambil seluruh data pada kolom Species, sehingga y berisi nama species bunga Iris seperti setosa, versicolor, dan virginica. Data pada y nantinya akan diubah menjadi bentuk angka menggunakan LabelEncoder agar dapat digunakan dalam proses klasifikasi

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Pada code ini digunakan untuk mengubah label species yang sebelumnya berupa teks menjadi bentuk angka agar dapat digunakan dalam proses klasifikasi. LabelEncoder digunakan untuk membuat encoder, kemudian lab.fit_transform(y) digunakan untuk mempelajari dan mengubah nilai pada y menjadi angka. Hasilnya adalah setosa menjadi 0, versicolor menjadi 1, dan virginica menjadi 2. Setelah itu, y digunakan untuk menampilkan hasil perubahan label tersebut dalam bentuk array angka.

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

Pada code ini digunakan untuk membagi dataset menjadi data training dan data testing yang akan digunakan dalam proses klasifikasi. train_test_split digunakan untuk melakukan pembagian data, kemudian X dibagi menjadi X_train sebagai data fitur untuk training dan X_test sebagai data fitur untuk testing. Begitu juga dengan y, yang dibagi menjadi y_train sebagai label training dan y_test sebagai label testing. test_size=0.2 berarti 20% data digunakan untuk testing dan 80% untuk training, sedangkan random_state=0 digunakan agar pembagian data tetap sama setiap kali kode dijalankan.

In [6]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)


Pada code ini digunakan untuk mengimpor dan membuat model Gaussian Naïve Bayes yang digunakan untuk melakukan klasifikasi pada dataset Iris. GaussianNB() digunakan untuk membuat model Naive Bayes dengan distribusi Gaussian, sedangkan model.fit(X_train, y_train) digunakan untuk melatih model menggunakan data fitur X_train dan label y_train. Hasil model yang telah dilatih kemudian disimpan ke dalam variabel NB_model sehingga dapat digunakan untuk melakukan prediksi pada data testing maupun data baru.

In [7]:
result = NB_model.predict(X_test)
result


array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Pada code ini digunakan untuk melakukan prediksi terhadap data testing menggunakan model Gaussian Naïve Bayes yang sebelumnya sudah dilatih. NB_model.predict(X_test) digunakan untuk memprediksi species berdasarkan data fitur pada X_test, kemudian hasil prediksi disimpan ke dalam variabel result. Setelah itu, result digunakan untuk menampilkan hasil prediksi dalam bentuk angka, yaitu 0 untuk setosa, 1 untuk versicolor, dan 2 untuk virginica

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Pada code ini digunakan untuk menghitung tingkat akurasi model dalam melakukan klasifikasi pada data testing. accuracy_score(y_test, result) digunakan untuk membandingkan label sebenarnya pada y_test dengan hasil prediksi pada result. Hasil akurasi kemudian dikalikan 100 agar ditampilkan dalam bentuk persentase. Berdasarkan hasil pada modul, model menghasilkan accuracy sebesar 96.66666666666667

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\sara rajagukguk\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2830: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Pada code ini digunakan untuk melakukan prediksi terhadap data baru menggunakan model Gaussian Naive Bayes yang telah dilatih. input_data berisi nilai fitur baru yaitu [5.1, 3.5, 1.4, 0.2], kemudian NB_model.predict(input_data) digunakan untuk menentukan species berdasarkan data tersebut dan hasilnya disimpan pada result. Selanjutnya, lab.inverse_transform(result) digunakan untuk mengubah hasil prediksi yang masih berupa angka menjadi nama species. Terakhir, print() digunakan untuk menampilkan hasil prediksi, yaitu data tersebut diprediksi sebagai setosa.

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)


Hasil prediksi untuk input baru adalah: ['setosa']


Pada code ini digunakan untuk melakukan prediksi terhadap data baru dengan format DataFrame agar memiliki nama kolom yang sama dengan dataset saat model dilatih. pd.DataFrame() digunakan untuk membuat data baru [5.1, 3.5, 1.4, 0.2] dengan nama kolom dari iris.feature_names. Selanjutnya, NB_model.predict(input_data) digunakan untuk memprediksi species dari data tersebut, kemudian lab.inverse_transform(result) digunakan untuk mengubah hasil prediksi yang berupa angka menjadi nama species. Terakhir, print() digunakan untuk menampilkan hasil prediksi, yaitu setosa. Cara ini juga digunakan untuk menghindari warning karena nama fitur pada data input sudah sesuai dengan nama fitur yang digunakan saat model dilatih.

In [17]:
#TUGAS
input_data = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]),columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['virginica']


Pada code ini digunakan untuk prediksi data baru masuk ke spesies. pd.DataFrame() digunakan untuk membuat data baru [7.3, 3.8, 4.6, 2.3] dalam bentuk DataFrame dengan nama kolom yang sama seperti dataset Iris. Selanjutnya, NB_model.predict(input_data) digunakan untuk memprediksi species berdasarkan empat nilai fitur tersebut. Hasil prediksi yang masih berupa angka kemudian diubah menjadi nama species menggunakan lab.inverse_transform(result). Terakhir, print() digunakan untuk menampilkan hasil prediksi, yaitu virginica.